### ISSUE TITLE
WhichKeyVisual and commands that take range.

### ISSUE THREAD

**Issue by YaN-3k** [liuchengxu/vim-which-key#125]
**Environment information**
 - OS: linux
 - (Neo)Vim version: v0.4.3
 - vim-which-key version: 786dceaab0987e164dcee125b8ff6111851df625

**Bug description**
WhichKeyVisual does not work correctly with commands that take range.

**To Reproduce**

1. Create the minimal vimrc `min.vim` and `test_file`:
`min.vim`
```vim
call plug#begin('~/.config/nvim/bundle/')

Plug 'tpope/vim-commentary'
Plug 'liuchengxu/vim-which-key'

call plug#end()

" map leader
let mapleader = ' '

" map leader to which_key
nnoremap <silent> <leader> :silent <c-u> :silent WhichKey '<Space>'<CR>
vnoremap <silent> <leader> :silent <c-u> :silent WhichKeyVisual '<Space>'<CR>

" create map to add keys to
let g:which_key_map =  {}
" fast up
set timeoutlen=500

" `:Commentary accept range in visual mode`
let g:which_key_map['/'] = [ ':Commentary'  ,             'comment' ]

" register dictionary
call which_key#register('<Space>', "g:which_key_map")
```
`test_file`
```
1 test line
2 test line
3 test line
4 test line
5 test line
6 test line
```

2. Installl [vim-commenntary](https://github.com/tpope/vim-commentary) and WhichKey

3. Start (neo)vim with command: `vim -u min.vim test_file`

4. Type `V` and select more that one line

5. Type `gc` and see that many lines have been commented.

6. Repeat step 4, but with other lines. (or type undo)

7. Type `<space>/`

**Curren behavior**
Only the last selected line will be commented.

**Expected behavior**
All selected line will be commented.

**Video**
[demo](http://0x0.st/iOy_.mp4)

**Comment by cxgslegend** [comment]
I am having the same issue with commentary. It only comments out the last line.

**Comment by rene-descartes2021** [comment]
I have tried this minimal vimrc and it is still the case where only the last line is commented.

For workaround, change:
```vim
let g:which_key_map['/'] = [ ':Commentary'  ,             'comment' ]
```
To:
```vim
let g:which_key_map['/'] = [ '<Plug>Commentary'  ,             'comment' ]
```
And all lines will be commented.

I will look into why `:Commentary` does not work as expected.

**Comment by rene-descartes2021** [comment]
I took a look at supporting the `:Commentary` form with range. Just writing some thoughts on implementation details.

I can't find a way to get `<line1>` and `<line2>` to represent the visual range in
```vim
command! -bang -nargs=1 -range WhichKeyVisual call which_key#start(1, <bang>0, <args>, <line1>, <line2>)
```
They seem to be stuck on the last line of the visual range.

I've tried using many, not all combinations of `<Plug>`, `<Cmd>`, and `<expr>`, e.g.:
```vim
" plugin/which_key.vim
xnoremap <Plug>(which_key_visual) <Cmd>WhichKeyVisual nr2char(getchar())<CR>
```
```vim
" end-user vimrc
xmap <leader> <Plug>(which_key_visual)<Space>
```
No working combination I tried results in `<line1>` and `<line2>` not being on the last line. I'm sure there is a solution.

To support the `:Commentary` form, I suppose that a check could be made [here](https://github.com/liuchengxu/vim-which-key/blob/9875baefb1d3e78abbf8f652a17e130744a37742/autoload/which_key.vim#L375), e.g.:
```vim
if s:vis
  let Cmd = line('v').','.line('.').Cmd
endif
```
That would probably work best when considering command arguments, as AFAIK no form of `<Plug>` would pass arguments, as e.g. `:Commentary Foo Bar` has no workaround `<Plug>` form, and thus is an unsatisfactory workaround in general. I'll make PR with this.
